ui/src/app/machines/views/MachineDetails/MachineTests/MachineTestsTable/MachineTestsTable.tsx

```diff
@@ -16,19 +20,183 @@ const isSuppressible = (result: ScriptResult) =>
   result.status === scriptStatus.TIMEDOUT ||
   result.status === scriptStatus.FAILED_APPLYING_NETCONF;
 
+const renderActions = (
+  result: ScriptResult,
+  showPreviousTests: (id: ScriptResult["id"]) => void,
+  hidePreviousTests: (id: ScriptResult["id"]) => void,
+  hasHistory: boolean,
+  isExpanded: boolean
+) => {
+  const links = [];
+  if (!hasHistory) {
+    return null;
+  }
+  if (!isExpanded) {
+    links.push({
+      children: "View previous tests",
+      onClick: () => showPreviousTests(result.id),
+      "data-test": "action-menu-show-previous",
+    });
+  }
+  if (isExpanded) {
+    links.push({
+      children: "Hide previous tests",
+      onClick: () => hidePreviousTests(result.id),
+      "data-test": "action-menu-hide-previous",
+    });
+  }
+  return (
+    <TableMenu
+      data-test="action-menu"
+      links={links}
+      position="right"
+      title="Take action:"
+    />
+  );
+};
+
 const MachineTestsTable = ({
   machineId,
   scriptResults,
 }: Props): JSX.Element => {
   const dispatch = useDispatch();
+  const [visibleHistory, setVisibleHistory] = useState([]);
+
+  const history = useSelector(scriptResultSelectors.history);
 
   const showPreviousTests = (id: ScriptResult["id"]) => {
-    dispatch(scriptResultActions.getHistory(id));
+    setVisibleHistory([...new Set([...visibleHistory, id])]);
   };
 
+  const hidePreviousTests = (id: ScriptResult["id"]) => {
+    setVisibleHistory([
+      ...new Set(visibleHistory.filter((visibleId) => visibleId !== id)),
+    ]);
+  };
+
+  useEffect(() => {
+    if (Object.keys(history).every((k) => !history[k].length)) {
```

Comment:
For the sake of readability, this could probably be turned into a variable.
``` typescript
const noHistory = Object.keys(history).every((k) => !history[k].length));
if (noHistory) {
...
```